File: backend/src/main/java/org/luxons/sevenwonders/game/effects/Effect.java
package org.luxons.sevenwonders.game.effects;

import org.luxons.sevenwonders.game.api.Table;

/**
 * Represents an effect than can be applied to a player's board when playing a card or building his wonder. The effect
 * may affect (or depend on) the adjacent boards. It can have an instantaneous effect on the board, or be postponed to
 * the end of game where point calculations take place.
 */
public interface Effect {

    void apply(Table table, int playerIndex);

    int computePoints(Table table, int playerIndex);
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/effects/Effect.kt
package org.luxons.sevenwonders.game.effects

import org.luxons.sevenwonders.game.Player

/**
 * Represents an effect than can be applied to a player's board when playing a card or building his wonder. The effect
 * may affect (or depend on) the adjacent boards. It can have an instantaneous effect on the board, or be postponed to
 * the end of game where point calculations take place.
 */
interface Effect {

    fun applyTo(player: Player)

    fun computePoints(player: Player): Int
}
